src/game.zig

```diff
@@ -277,6 +278,60 @@ pub const collision = struct {
 		return resultBox;
 	}
 
+	pub fn calculateFriction(comptime side: main.utils.Side, pos: Vec3d, hitBox: Box, defaultFriction: f32) f32 {
+		const boundingBox: Box = .{
+			.min = pos + hitBox.min,
+			.max = pos + hitBox.max,
+		};
+		const minX: i32 = @intFromFloat(@floor(boundingBox.min[0]));
+		const maxX: i32 = @intFromFloat(@floor(boundingBox.max[0] - 0.0001));
+		const minY: i32 = @intFromFloat(@floor(boundingBox.min[1]));
+		const maxY: i32 = @intFromFloat(@floor(boundingBox.max[1] - 0.0001));
+
+		const z: i32 = @intFromFloat(@floor(boundingBox.min[2] - 0.01));
+
+		var friction: f64 = 0;
+		var totalArea: f64 = 0;
+
+		var x = minX;
+		while (x <= maxX) : (x += 1) {
+			var y = minY;
+			while (y <= maxY) : (y += 1) {
+				const _block = if(side == .client) main.renderer.mesh_storage.getBlock(x, y, z)
+				else main.server.world.?.getBlock(x, y, z);
+
+				if (_block) |block| {
+					const blockPos: Vec3d = .{@floatFromInt(x), @floatFromInt(y), @floatFromInt(z)};
+
+					const blockBox: Box = .{
+						.min = blockPos + @as(Vec3d, @floatCast(main.models.models.items[block.mode().model(block)].min)),
+						.max = blockPos + @as(Vec3d, @floatCast(main.models.models.items[block.mode().model(block)].max)),
+					};
+
+					if (boundingBox.min[2] > blockBox.max[2] or boundingBox.max[2] < blockBox.min[2]) {
+						continue;
+					}
+
+					const min = @max(vec.xy(blockBox.min), vec.xy(boundingBox.min));
+					const max = @min(vec.xy(blockBox.max), vec.xy(boundingBox.max));
```

Comment:
Actually I think this needs to be a full clamp, otherwise we may end up with max being smaller than min.